### ISSUE TITLE
Export FUSE mount from batch FTAs

### ISSUE THREAD

**Issue by brettkettering** [mar-file-system/marfs#42]
Work with Chris Mitchell and team to setup the batch FTAs to re-export the FUSE mount.

This allows us not to worry about unsharing the MarFS GPFS metadata file system mount. And, it allows us to IPtables deny the interactive FTAs to access the object store.

**Comment by brettkettering** [comment]
Use the pcpqa account to test that this works for a normal user. Jeff Inman is here to try this.

Plan A: We think a normal NFS mount works RW, but need to verify it with a normal user.

Plan B: sshfs on the interactive FTA works with a RW re-export.

GPFSt can mount RO with a NFS mount.

**Comment by brettkettering** [comment]
I wanted to capture Gary's thoughts on ways to avoid having to worry about authentication, with our without SSL.

Here's his input:
Eventually if we are worried about escalation (which I think all security issues boil down to) on the interactive FTA’s, we can add containers/jails, or even more secure concepts. We could use a forwarding FUSE daemon and forward all MarFS and other file system requests over to the batch FTAs. We could even use a private FUSE daemon to each user and or process family. This would enable us not to have the password stored on the interactive FTAs at all, so root escalation there would buy you nothing, and sniffing for passwords would be useless.  Further if we did this forwarding thing we could make it so that root escalation on the interactive FTAs doesn’t allow you to get onto the storage servers (different root creds and other protections) along with jails and containers and even MLS. We could trivially do MLS on these interactive FTAs because they don’t mount anything directly, it's all done via forwarding, making root escalation extremely difficult and even more useless.  We wouldn’t use MLS on the batch FTAs or storage servers because of support issues, but users cant get onto those ,anyway. This solves almost fundamentally the root escalation problem for the FTAs, not just for MarFS but for all storage the FTAs touch. Yes, I know it doesn’t solve the storage admin can see anything problem, but it stops the stupid red team trivial case of, "I became root on a user facing machine and captured the flag" issue.  It would be lovely to say to the red team, "So what? You became root. What now?"

Having the concept of interactive and batch FTAs combined with forwarding is a very powerful concept, that we have only begun to exploit.  It is essentially a glove box and if we control the forwarding layer. It’s a glove box we control.

**Comment by jti-lanl** [comment]
Just for the record, I believe our alternative approach accomplishes the same end:
only mounting fuse on one (or more) of the batch FTAs, and then exporting it via
NFS to the interactive FTA(s).  Again, no creds are stored on the interactive node,
so gaining root there doesn't buy "universal access".

However, as root, I'd think you could assume any given user's identity and
read files as them -- probably true of any file system.  So, I guess we do still have to
care about people getting root on the interactive FTA, but only for the usual reasons.

Jeff

**Comment by brettkettering** [comment]
At this time we are able to look at files on the exported NFS mount. We will try reading and writing files and try with big files.

**Comment by jti-lanl** [comment]
Current status:  basic functionality (write/read a small file) appears to work, but there are some issues with how the user ID gets propagated through fuse.  Investigation lead to issue #77, where we were screwing up setting the euid in fuse.  Experimental approach using setfsuid() seems to have resolved that, but I'm still testing.

Maybe it's also a problem that pcpqa exists only on the client, and not on the server, in the ODSU test environment.  Switching to the ccstar cluster, where I can have the same user on both sides.

Also, getting a better picture of trash-directory permission requirements.

**Comment by jti-lanl** [comment]
See issues in #79.

**Comment by brettkettering** [comment]
There are issues with both remount solutions.

1) NFS mounted MarFS in RW has issues with multiple clients accessing MarFS through one mount point. We think the ultimate solution to this is to implement SEMI_DIRECT with an automated backend script that moves things from a mostly POSIX compliant underlying file system (like Lustre, for example) to the object store. It will take longer to implement this solution than we have time for the start of Open Science. It is on the high priority list for post-Open Science and targeted at secure Trinity Campaign Storage (right now that target start date is 7/1/16, so we need to be ready about 2-4 weeks before that.

2) SSHFS mounted MarFS in RW has issues with needing a mount point per user on the interactive FTAs. The FTA team thinks this will cause issues. Furthermore, it requires sftp, which then a user can execute and target the implementation-specific GPFS file system that is the metadata file system. The Admin team is looking at ways to limit the SSH commands that can be sent and what they can target.

3) NFS mounted MarFS in RO does not allow users to make directories and change permissions, groups, etc. Users need to do this to share files amongst team members.

So, for now we are going to implement the username/password capability for MarFS FUSE, PFTool, and whatever batch scripts interact with the object storage.

**Comment by brettkettering** [comment]
From Dave Bonnie:

To clarify #1 - this issues lie in any users (single or multiple) writing, since NFS is trying to be tricky and write things out of order.  We’re investigating to see if there’s a good solution to this, but it is not being relied upon for open science.

**Comment by jti-lanl** [comment]
I think Brett's (3), is referring to mounting RO in the NFS sense.  However, we could also consider using the configuration iperms/bperms to remove only the WD property from the affected Repos. That would currently disable the following ops:

* truncate/ftruncate
* mknod / open(O_CREAT)
* open(O_WRONLY)      [O_RDWR is forbidden]
* unlink
* write

That leaves in place chmod, chown, mkdir, etc,  though rename would fail, I think.

**Comment by brettkettering** [comment]
An update...

We have to do these things:

1) Mount all the other file systems in the chroot environment.

2) Confirm that the pfcp, pfls, and pfcm commands can still be executed properly in the chroot environment. It's important to note that the words “the chroot environment” may not be completely accurate.

What we want on the server side:

a) 2 environments, one for sftp/scp (which are native sshd operations I guess, not fork and exec like pfcp pfls etc.). This must hide the gpfs mdfs so sshfs can wonder into the mdfs, thus the sftp only stanza in sshd that limits what sftp/scp/sshfs can see there would be a second environment that is active when you do a fork/exec out of sshd – for pftp pfls etc. This environment may or may not be chrooted (its up to the admins) but this environment must be able to see all the mdfs's and have access to the secret password in a protected way and of course it must limit what the user can run to only a set of controlled things on the client (interactive fta) side I assume you are considering using chroot to keep everyones namespace to be separate so each can have its own mounts etc.

 So there are multiple potential uses for chroot. There are multiple environments (3 I think): the interactive fta env set up to allow users to have their own mounts via private namespaces/chroot, however you want to do this; the env on the server for sftp/scp operations to execute, presumably done via sshd config on the server side to limit sftp/scp to not see mdfs/password stash; and the env on the server for pfcp/pfls to execute, may or may not want to chroot or protect this, but this env has to see mdfs and password stash (protected by setuid or something),  and must limit severely what can be run (what sshd can fork/exec). Being careful that we don’t think there is just one environment and it's chrooted…

3) Confirm that even though when you “ls” the chroot “/“ you don’t see the “/gpfs/...” metadata file system, you still can’t sftp or scp a file from the interactive FTA onto or off of that /gpfs/... file system.

4) Write a script that runs periodically and reconciles the users in the password file with the /marfs/$USER user directories and ensures that each of those /marfs/$USER directories is mounted as $USER (not root) using SSHFS. It will update the /etc/fstab file and do the proper mounts and umounts to ensure the proper SSHFS file systems are mounted. Need to test multiple users mounting the same thing on one (the same) interactive fta at the same time, just to ensure that you have indeed created a multi-user interactive fta and you have killed any worry about name collisions in mount space.

5) Change the interactive and batch FTAs to implement the SSHFS and chroot environment.

**Comment by brettkettering** [comment]
Manually mounting the SSHFS works. Still need to workout the CFengine configuration process.

**Comment by brettkettering** [comment]
Decided to utilize FUSE for Interactive node for Open Science. There are too many complications trying to get SSHFS to work with the systems.  It could probably work, but we have complications with multiple mounts on the interactive node, and the changes with pftool on the bash FTAs.  I reviewed this with Alex this morning to bring him up to speed on the decision.  We probably need to review this at the CCB this week.

**Comment by brettkettering** [comment]
Even using SEMI_DIRECT, read performance is poor. We need to investigate why the read performance is poor and how we may be able to improve it.

**Comment by brettkettering** [comment]
We may decide not to do this and just have the MarFS FUSE daemon mount on the interactive FTAs in an unshare environment so that the metadata implementation is not exposed to the user.

**Comment by brettkettering** [comment]
The benchmark of success is 100MB/s transfer rates. It does functionally work, but need to do the performance test.

Testing with the FUSE daemon running and mounted on the interactive FTA shows transfer rates at 100MB/s.

Testing with the FUSE daemon running on the batch FTA and then NFS mounted on the interactive FTA shows transfer rates at 10MB/s.

Will test directly on the batch FTA and see what it is there.

**Comment by brettkettering** [comment]
We have seen that NFS thinks its block size is 1MB. FUSE thinks its block size is 256KB. We need to put FUSE into debug mode on the batch FTA and figure out what size blocks NFS is asking FUSE to get. We also want to see if there are requests from multiple threads on the same file handle, which MarFS FUSE will use locks to control, and consequently slow down performance.

We may need to figure a way to get bigger block requests from NFS to FUSE and handle concurrent requests to the same file in parallel rather than sequentializing them with locks.

**Comment by jti-lanl** [comment]
nfsd submits read-sizes of 4KB.   Furthermore, within any given thread, the read-offsets are typically not contiguous (i.e. different nfsd threads must be submitting requests for different 4K chunks of a contiguous stretch -- it's hard to tell, because what we see are multiple fuse threads).  It may be that nfsd thinks this is how you maximize performance for reads of large blocks of data: imagining that it is keeping up the queue-depth on some underlying disks.

There was speculation that NFS might be looking at the root dir (under which fuse was mounted on the server-side), and getting distracted by that block size, which in the case of a root that was mounted on tmpfs might be small.   However, mounting the server-side fuse under GPFS (e.g. /gpfs/marfs/fuse_mnt) doesn't change the 4K read-sizes seen by fuse.

marfs_fuse receives a statvfs call at export time, and another at remote mount time, and it responds in both cases with a statbuf gotten from an internal statvfs call to the MDFS, which in our case specifies a block-size of 256K.

--- Possible solutions/improvement:

* marfs_read() queues out-of-order read-requests, as an experimental attempt to allow reads from other threads to bring them into order.  I'm not convinced this works as hoped, and it could just have the effect of introducing delays, if the order of requests from nfsd has certain properties.  We should try just subverting all the fanciness: don't attempt to detect multi-threading, don't enqueue out-of-order requests, just go ahead an do the close/re-open

* do read-aheads in the case of out-of-order requests (e.g. read 256KB), then keep the buffers around, referenced by their offset and length.  Look through the buffer-list to satisfy out-of-order requests.  If the buffer-list gets too big, deallocate old buffers, and replace with new ones.  This could have the effect of doing some extra reading, but may also cut down on close/re-opens, and may allow many small requests to be satisfied from the read-ahead buffers.

**Comment by cadejager** [comment]
I had been trying to test the mount with an rsize and wsize of 4M. If we reduce that to 4K might it serialize the requests?

**Comment by jti-lanl** [comment]
It’s worth a try.  There are also parameters to control read-ahead, etc, that might make a difference.

**Comment by brettkettering** [comment]
Let's try something that we won't do in production, but will confirm that this issue is due to nfsd and using the NFS mount of the FUSE on the batch FTA. Using the same FUSE on the batch FTA directly from the batch FTA, do the same read you are doing when you see 4KB requests and tell us what the size of the requests are in this mode. If our hypothesis from yesterday is correct, it should be 128KB. By any means if it is 128KB or 256KB, and not 4KB, we at least know that the problem is with nfsd because we're using the same FUSE in both cases, one time with nfsd and the other without.

If it is confirmed to be nfsd, what does the blogosphere say about FUSE and nfsd, and getting bigger block sizes to work?

**Comment by brettkettering** [comment]
Jeff I.: Did we try this experiment and determine what the block size is when doing the exact same operation directly on the batch FTA to the FUSE mount instead of using NFS on the interactive FTA to get to it?

Dave B.: You mentioned that there was some setup you knew about that you thought would change nfsd doing 4KB reads when used from the interactive FTA. What specifically is that? How do we set it up to try it?

Thanks,
Brett

**Comment by thewacokid** [comment]
Now that we have the pftool batch performance somewhat sorted out, I’ll start taking a look at this.  The issues we had last week with the storage nodes likely impacted the NFS testing, but not enough to cover up 4KB IOs if that’s what is being issued.

- Dave

**Comment by jti-lanl** [comment]
Yes, I saw nfsd issuing nothing but 4K reads to fuse.  I’m open to trying options at any of the mount-stages, if you think they will affect this.

Meanwhile, I told Brett I would repeat the tests, comparing the same fuse-mount, (a) through NFS, and (b) directly from the local host, to confirm whether they are behaving differently.

Thanks,
Jeff

**Comment by thewacokid** [comment]
Was this with NFS v3 or v4?  For the re-export, it’s probably easier to get v3 to behave properly.  NFS v4 checks the export paths differently for the FS block size.  I’ll try to get a few pointed tests in today to suss this out.

**Comment by jti-lanl** [comment]
It was on NFS v4.  (But I may have moved to v4 because I was seeing the same thing on v3.  Not positive about that.)

**Comment by brettkettering** [comment]
Turned-off directio on the FUSE mount and used NFSv3. Now we get 60MB/s. We think if we put n+e threads under FUSE in the MC DAL we'll be able to read faster because we can be reading from all the MC file systems concurrently.
